### PR TITLE
Be more permisive with comment bodies

### PR DESCRIPTION
**Pull request by davidcornu**
Fixes https://github.com/Shopify/liquid/issues/936

**Changes**

- Adds a subclass of `BlockBody` called `CommentBody` which re-implements `parse` to ignore anything that isn't a comment tag
- Updates `Comment` to use `CommentBody` instead of `BlockBody`

### CODE REVIEW

File: test/integration/tags/standard_tag_test.rb [hunk 1]
```diff
@@ -31,9 +31,10 @@ def test_has_a_block_which_does_nothing
 
     assert_template_result('', '{%comment%}{%blabla%}{%endcomment%}')
     assert_template_result('', '{% comment %}{% blabla %}{% endcomment %}')
+    assert_template_result('', '{% comment %}{% if "foo" %}{% endcomment %}')
```
Comment (davidcornu) [inline test/integration/tags/standard_tag_test.rb]: Confirmed this breaks on master

```
~/Development/liquid ignore-comment-body
❯ git checkout master -- lib/liquid/tags/comment.rb

~/Development/liquid ignore-comment-body*
❯ b rake test
-- LAX ERROR MODE
Run options: --seed 39200

# Running:

..........................................................................................................................................................................................................................................................................................................................................................................................................................E........................................................................................................................................................

Finished in 0.158554s, 3550.8407 runs/s, 8640.5893 assertions/s.

  1) Error:
StandardTagTest#test_has_a_block_which_does_nothing:
Liquid::SyntaxError: Liquid syntax error: Unknown tag 'endcomment'
    /Users/david/Development/liquid/lib/liquid/block.rb:38:in `unknown_tag'
    /Users/david/Development/liquid/lib/liquid/tags/if.rb:36:in `unknown_tag'
    /Users/david/Development/liquid/lib/liquid/block.rb:68:in `block in parse_body'
    /Users/david/Development/liquid/lib/liquid/block_body.rb:33:in `parse'
    /Users/david/Development/liquid/lib/liquid/block.rb:58:in `parse_body'
    /Users/david/Development/liquid/lib/liquid/tags/if.rb:24:in `parse'
    /Users/david/Development/liquid/lib/liquid/tag.rb:10:in `parse'
    /Users/david/Development/liquid/lib/liquid/block_body.rb:27:in `parse'
    /Users/david/Development/liquid/lib/liquid/block.rb:58:in `parse_body'
    /Users/david/Development/liquid/lib/liquid/block.rb:12:in `parse'
    /Users/david/Development/liquid/lib/liquid/tag.rb:10:in `parse'
    /Users/david/Development/liquid/lib/liquid/block_body.rb:27:in `parse'
    /Users/david/Development/liquid/lib/liquid/document.rb:10:in `parse'
    /Users/david/Development/liquid/lib/liquid/document.rb:5:in `parse'
    /Users/david/Development/liquid/lib/liquid/template.rb:132:in `parse'
    /Users/david/Development/liquid/lib/liquid/template.rb:116:in `parse'
    /Users/david/Development/liquid/test/test_helper.rb:40:in `assert_template_result'
    /Users/david/Development/liquid/test/integration/tags/standard_tag_test.rb:34:in `test_has_a_block_which_does_nothing'

563 runs, 1370 assertions, 0 failures, 1 errors, 0 skips
```

File: test/integration/tags/standard_tag_test.rb [hunk 2]
```diff
@@ -31,9 +31,10 @@ def test_has_a_block_which_does_nothing
...
     assert_template_result('', '{%comment%}{% endif %}{%endcomment%}')
     assert_template_result('', '{% comment %}{% endwhatever %}{% endcomment %}')
-    assert_template_result('', '{% comment %}{% raw %} {{%%%%}}  }} { {% endcomment %} {% comment {% endraw %} {% endcomment %}')
+    assert_template_result(' ', '{% comment %}{% raw %} {{%%%%}}  }} { {% endcomment %} {% comment {% endraw %} {% endcomment %}')
```
Comment (davidcornu) [inline test/integration/tags/standard_tag_test.rb]: For reasons I have yet to discover, `CommentBody` parses this as

1. `"{% comment %}{% raw %} {{%%%%}}  }} { {% endcomment %}"`
2. `" "`
3. `"{% comment {% endraw %} {% endcomment %}"`

whereas `BlockBody` treats the whole thing as a single comment.

Comment (davidcornu) [reply test/integration/tags/standard_tag_test.rb]: Found it. For `BlockBody`, `{% raw %}` swallows up the first `{% endcomment %}`.
```
"{% comment %}"
    "{% raw %}"
        " {{%%%%}}  }} { {% endcomment %} {% comment "
    "{% endraw %}"
    " "
"{% endcomment %}"
```

File: lib/liquid/tags/comment.rb [hunk 3]
```diff
@@ -10,6 +10,12 @@ def unknown_tag(_tag, _markup, _tokens)
     def blank?
       true
     end
+
+    def parse(tokens)
+      @body = CommentBody.new
```
Comment (davidcornu) [inline lib/liquid/tags/comment.rb]: Another approach would be to do the same as `Liquid::Raw#parse`

https://github.com/Shopify/liquid/blob/27c91203ab4adfde2cc5e1142bce2a267634ef8e/lib/liquid/tags/raw.rb#L12-L23

but it wouldn't support things like `{% comment %}{% comment %}{% endcomment %}{% endcomment %}`

```
~/Development/liquid master
❯ irb -I ./lib
irb(main):001:0> require 'liquid'
=> true
irb(main):002:0> Liquid::Template.parse('{% raw %}{% raw %}{% endraw %}{% endraw %}')
Liquid::SyntaxError: Liquid syntax error: Unknown tag 'endraw'
	from /Users/david/Development/liquid/lib/liquid/document.rb:23:in `unknown_tag'
	from /Users/david/Development/liquid/lib/liquid/document.rb:11:in `block in parse'
	from /Users/david/Development/liquid/lib/liquid/block_body.rb:33:in `parse'
	from /Users/david/Development/liquid/lib/liquid/document.rb:10:in `parse'
	from /Users/david/Development/liquid/lib/liquid/document.rb:5:in `parse'
	from /Users/david/Development/liquid/lib/liquid/template.rb:132:in `parse'
	from /Users/david/Development/liquid/lib/liquid/template.rb:116:in `parse'
	from (irb):2
	from /usr/local/bin/irb:11:in `<main>'
```

**Comment by davidcornu** [comment]
@dylanahsmith @pushrax would love some feedback on the implementation.

CI is angry at me because of linter errors (I re-used `BlockBody#parse`'s structure which has more than 3 levels of block nesting), the the rest of the test suite runs successfully. I'd be glad to refactor but this is only an exploration at this point.

File: test/integration/tags/standard_tag_test.rb (hunk 2 above)
Comment (pushrax) [reply test/integration/tags/standard_tag_test.rb]: I feel like this workaround (raw in comment) could be pretty common and that this change will subtly break some uses of `{% comment %}` :(

Comment (pushrax) [reply test/integration/tags/standard_tag_test.rb]: cc @fw42 ^ if this test breaks how bad would it be?

**Comment by pushrax** [comment]
Could this be implemented by extending `Raw`?

**Comment by davidcornu** [comment]
> Could this be implemented by extending `Raw`?

`{% raw %}` doesn't support nesting - https://github.com/Shopify/liquid/pull/937#discussion_r144982950

File: lib/liquid/comment_body.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+            whitespace_handler(token, parse_context)
+            if token =~ FullToken
+              tag_name = $1
+              markup = $2
+
+              if tag_name == 'comment'
+                new_tag = Comment.parse(tag_name, markup, tokenizer, parse_context)
+                @blank &&= new_tag.blank?
+                @nodelist << new_tag
+              else
```
Comment (dylanahsmith) [inline lib/liquid/comment_body.rb]: I think you need to handle the `raw` tag as well, given that it was previously the method used to workaround syntax errors in comments.

File: lib/liquid/comment_body.rb [hunk 5]
```diff
@@ -0,0 +1,36 @@
+module Liquid
+  class CommentBody < BlockBody
+    def parse(tokenizer, parse_context)
+      parse_context.line_number = tokenizer.line_number
+      while token = tokenizer.shift
+        unless token.empty?
```
Comment (dylanahsmith) [inline lib/liquid/comment_body.rb]: This condition isn't necessary.

File: lib/liquid/comment_body.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+          if token.start_with?(TAGSTART)
+            whitespace_handler(token, parse_context)
+            if token =~ FullToken
+              tag_name = $1
+              markup = $2
+
+              if tag_name == 'comment'
+                new_tag = Comment.parse(tag_name, markup, tokenizer, parse_context)
+                @blank &&= new_tag.blank?
+                @nodelist << new_tag
```
Comment (dylanahsmith) [inline lib/liquid/comment_body.rb]: We don't need this nested comment tag in the nodelist either.

**Comment by ADTC** [comment]
Any update on this?

**Review by devdreamsolution — APPROVED** [review]

**Comment by davidcornu** [comment]
Didn't realize this PR is still open. Ended up giving up on this particular approach as it could easily end up being a breaking change due to the `raw` tag's current behaviour (https://github.com/Shopify/liquid/pull/937#discussion_r145189104).

There's likely a solution that takes that into consideration but I lack the depth of knowledge of Liquid's parsing logic to pull that off safely.

**Pull request closed without merging**